ui/app/components/auth-jwt.js

```diff
@@ -16,200 +18,226 @@ const ERROR_MISSING_PARAMS =
+      this.watchCurrent.perform(oidcWindow);
+      // wait for message posted from oidc callback
+      // see issue https://github.com/hashicorp/vault/issues/12436
+      // ensure that postMessage event is from expected source
+      let event;
+      while (!event) {
+        // continue to wait for the correct message
+        event = await waitForEvent(thisWindow, 'message');
+      }
+
       if (event.origin === thisWindow.origin && event.isTrusted && event.data.source === 'oidc-callback') {
         return this.exchangeOIDC.perform(event.data, oidcWindow);
       }
-      // continue to wait for the correct message
-    }
-  }),
-
-  watchPopup: task(function* (oidcWindow) {
-    while (true) {
-      const WAIT_TIME = Ember.testing ? 50 : 500;
-
-      yield timeout(WAIT_TIME);
-      if (!oidcWindow || oidcWindow.closed) {
-        return this.handleOIDCError(ERROR_WINDOW_CLOSED);
+    })
+  );
+
+  watchPopup = task(
+    waitFor(async (oidcWindow) => {
+      while (oidcWindow && !oidcWindow.closed) {
+        const WAIT_TIME = Ember.testing ? 50 : 500;
+        await timeout(WAIT_TIME);
       }
-    }
-  }),
-
-  watchCurrent: task(function* (oidcWindow) {
-    // when user is about to change pages, close the popup window
-    yield waitForEvent(this.getWindow(), 'beforeunload');
-    oidcWindow.close();
-  }),
-
-  exchangeOIDC: task(function* (oidcState, oidcWindow) {
-    if (oidcState === null || oidcState === undefined) {
-      return;
-    }
-    this.onLoading(true);
-
-    let { namespace, path, state, code } = oidcState;
-
-    // The namespace can be either be passed as a query parameter, or be embedded
-    // in the state param in the format `<state_id>,ns=<namespace>`. So if
-    // `namespace` is empty, check for namespace in state as well.
-    if (namespace === '' || this.flagsService.hvdManagedNamespaceRoot) {
-      const i = state.indexOf(',ns=');
-      if (i >= 0) {
-        // ",ns=" is 4 characters
-        namespace = state.substring(i + 4);
-        state = state.substring(0, i);
+      return this.handleOIDCError(ERROR_WINDOW_CLOSED);
+    })
+  );
+
+  watchCurrent = task(
+    waitFor(async (oidcWindow) => {
+      // when user is about to change pages, close the popup window
+      await waitForEvent(this.getWindow(), 'beforeunload');
+      oidcWindow.close();
+    })
+  );
+
+  exchangeOIDC = task(
+    waitFor(async (oidcState, oidcWindow) => {
+      if (oidcState === null || oidcState === undefined) {
+        return;
+      }
+      this.args.onLoading(true);
+
+      let { namespace, path, state, code } = oidcState;
+
+      // The namespace can be either be passed as a query parameter, or be embedded
+      // in the state param in the format `<state_id>,ns=<namespace>`. So if
+      // `namespace` is empty, check for namespace in state as well.
+      // TODO smoke test HVD flag here and add test
+      if (namespace === '' || this.flags.hvdManagedNamespaceRoot) {
```

Comment:
the check for HVD was added by this PR: https://github.com/hashicorp/vault/pull/11283 but there's no test or deeper explanation as to why this fixed the bug (the changelog was copy pasta'd 🍝 ) 

I'll investigated and add a test in a follow-on PR